### PR TITLE
Implement storage fees and limits

### PR DESCRIPTION
**Pull request by MathieuDutSik**
## Motivation

We need to implement the storage fees for the linera-protocol. This version implements it in a rather straightforward way. We are counting the number of reads and writes, not the total data used.

## Proposal

* The `pricing.rs` module is significantly extended to separate writes from reads.
* The `remaining_fuel` is replaced by a `runtime_meter` that accounts for all the operation being done.
* As opposed to what we initially planned, all the computation is done in `runtime.rs`.
* As discussed beforehand, we also added limits and those limits are part of the `pricing.rs` code.

## Test Plan

CI so far, needs to extend.

## Release Plan

- Need to bump the major/minor version number in the next release of the crates.
- Need to update the developer manual.
- This PR is adding or removing Cargo features.
- Release is blocked and/or tracked by other issues (see links below)

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

**Review by ma2bd — COMMENTED** [review]
Just a quick comment in passing

File: linera-execution/src/lib.rs [hunk 1]
```diff
@@ -50,6 +50,44 @@ use serde::{Deserialize, Serialize};
...
+    /// The bytes that have been written
+    pub bytes_write: u64,
+    /// The maximum size of read allowed
+    pub maximum_bytes_read: u64,
+    /// The maximum size of write allowed
+    pub maximum_bytes_write: u64,
+}
+
+#[cfg(any(test, feature = "test"))]
+pub fn get_default_runtime_meter() -> RuntimeMeter {
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-execution/src/lib.rs:71]: Ok.

File: linera-execution/src/pricing.rs [hunk 2]
```diff
@@ -9,19 +9,42 @@ use serde::{Deserialize, Serialize};
 use thiserror::Error;
 
 /// A collection of costs associated with blocks in validators.
-#[derive(Eq, PartialEq, Hash, Clone, Debug, Default, Serialize, Deserialize, InputObject)]
+#[derive(Eq, PartialEq, Hash, Clone, Debug, Serialize, Deserialize, InputObject)]
 pub struct Pricing {
```
Comment (ma2bd) [inline linera-execution/src/pricing.rs:13]: Perhaps we could discuss a new name such as `Resource(Control)Policy` if this is going to include limits. @afck ?

File: linera-execution/src/runtime.rs [hunk 3]
```diff
@@ -33,12 +33,22 @@ use std::{
...
+    /// the number of read
+    n_read: Arc<AtomicU64>,
+    /// the total size being read
+    bytes_read: Arc<AtomicU64>,
+    /// the total size being write
+    bytes_write: Arc<AtomicU64>,
+    /// The maximum size of read allowed
+    maximum_bytes_read: u64,
+    /// The maximum size of write allowed
+    maximum_bytes_write: u64,
```
Comment (ma2bd) [inline linera-execution/src/runtime.rs:47]: Why not `Arc<RuntimeMeter>` where `RuntimeMeter` uses atomic integers?
Then you attach a number of (`&self`) update methods to `RuntimeMeter`

File: linera-execution/src/runtime.rs [hunk 4]
```diff
@@ -366,7 +386,7 @@ where
         match self
             .active_view_user_states_mut()
             .await
-            .get(&self.application_id())
+            .get_mut(&self.application_id())
```
Comment (ma2bd) [inline linera-execution/src/runtime.rs:389]: Is this needed?

File: linera-execution/src/runtime.rs [hunk 5]
```diff
@@ -381,9 +401,17 @@ where
...
```
Comment (ma2bd) [inline linera-execution/src/runtime.rs:404]: same question (and also below)

**Review by ma2bd — COMMENTED** [review]
As discussed we probably would like to decrease `remaining_fuel` continuously to exit early in case we exhaust the fuel (because of gas and/or storage fees).

File: linera-chain/src/chain.rs [hunk 6]
```diff
@@ -484,13 +484,31 @@ where
...
-        Self::sub_assign_fees(balance, pricing.storage_price(&block.incoming_messages)?)?;
-        Self::sub_assign_fees(balance, pricing.storage_price(&block.operations)?)?;
+        Self::sub_assign_fees(
+            balance,
+            pricing.storage_bytes_write_price(&block.incoming_messages)?,
+        )?;
+        Self::sub_assign_fees(
+            balance,
+            pricing.storage_bytes_write_price(&block.operations)?,
+        )?;
```
Comment (afck) [inline linera-chain/src/chain.rs:494]: I wonder whether we should keep these price categories separate, and just rename `storage_price` to `block_storage_price` or something?

An application's execution state will probably more often be read and overwritten on average, and sometimes deleted. On the other hand, there's no limit for how long it has to be kept.

Stored blocks can never be overwritten, but they might get archived after some time.

File: linera-execution/src/execution.rs [hunk 7]
```diff
@@ -232,10 +241,16 @@ where
...
         // Set the authenticated signer to be used in outgoing messages.
         result.authenticated_signer = signer;
-        *remaining_fuel = runtime.remaining_fuel();
+        *runtime_meter = runtime.runtime_meter();
+        if runtime_meter.bytes_read > runtime_meter.maximum_bytes_read {
+            return Err(ExecutionError::ExcessiveRead);
+        }
+        if runtime_meter.bytes_write > runtime_meter.maximum_bytes_write {
+            return Err(ExecutionError::ExcessiveWrite);
+        }
```
Comment (afck) [inline linera-execution/src/execution.rs:250]: We'll need to make sure that the maximum can't be exceeded (at least not by too much) within each call to `execute_operation` or `execute_message`, not just afterwards. (Could be a separate PR, of course.)

File: linera-execution/src/execution.rs [hunk 8]
```diff
@@ -232,10 +241,16 @@ where
...
         WASM_FUEL_USED_PER_BLOCK
             .with_label_values(&[])
-            .observe((initial_remaining_fuel - *remaining_fuel) as f64);
+            .observe((initial_remaining_fuel - runtime_meter.remaining_fuel) as f64);
```
Comment (afck) [inline linera-execution/src/execution.rs:253]: I just realized that this is wrong: This is the used fuel per user action, but there could be multiple (or zero) user actions in each block. (@andresilva91)

File: linera-execution/src/pricing.rs [hunk 9]
```diff
@@ -9,19 +9,42 @@ use serde::{Deserialize, Serialize};
...
     pub fuel: Amount,
-    /// The cost to store a block's operations and incoming messages, per byte.
-    pub storage: Amount,
+    /// The cost to read data per operation
+    pub storage_n_read: Amount,
+    /// The cost to read data per byte
+    pub storage_bytes_read: Amount,
+    /// The cost to store data per byte
+    pub storage_bytes_write: Amount,
+    /// The maximum data to read
```
Comment (afck) [inline linera-execution/src/pricing.rs:25]: "per block" (I think?)

File: linera-execution/src/runtime.rs [hunk 10]
```diff
@@ -485,6 +537,8 @@ where
     }
 
     async fn write_batch_and_unlock(&self, batch: Batch) -> Result<(), ExecutionError> {
+        let size = batch.size() as u64;
+        self.bytes_write.fetch_add(size, Ordering::Relaxed);
```
Comment (afck) [inline linera-execution/src/runtime.rs:541]: Should we compare to the write limit here, and return an error if it's too much?

File: linera-execution/src/runtime.rs (hunk 4 above)
Comment (MathieuDutSik) [reply linera-execution/src/runtime.rs:389]: No, corrected.

File: linera-execution/src/runtime.rs (hunk 5 above)
Comment (MathieuDutSik) [reply linera-execution/src/runtime.rs:404]: No, corrected.

File: linera-execution/src/pricing.rs (hunk 9 above)
Comment (MathieuDutSik) [reply linera-execution/src/pricing.rs:25]: Correct, very important point.

File: linera-execution/src/runtime.rs (hunk 10 above)
Comment (MathieuDutSik) [reply linera-execution/src/runtime.rs]: Yes, we can and I did it.

File: linera-execution/src/execution.rs [hunk 11]
```diff
@@ -232,10 +245,12 @@ where
         };
         // Set the authenticated signer to be used in outgoing messages.
         result.authenticated_signer = signer;
-        *remaining_fuel = runtime.remaining_fuel();
-        WASM_FUEL_USED_PER_BLOCK
-            .with_label_values(&[])
-            .observe((initial_remaining_fuel - *remaining_fuel) as f64);
+        let runtime_counts = runtime.runtime_counts();
+        let balance = self.system.balance.get_mut();
+        update_limits(balance, runtime_limits, pricing, runtime_counts)?;
```
Comment (afck) [inline linera-execution/src/execution.rs:250]: Should this be a method `update` of `RuntimeLimits`?

File: linera-execution/src/lib.rs [hunk 12]
```diff
@@ -50,6 +51,72 @@ use serde::{Deserialize, Serialize};
 use std::{io, path::Path, str::FromStr, sync::Arc};
 use thiserror::Error;
 
+/// substract an amount to a balence and report an error if that is impossible
```
Comment (afck) [inline linera-execution/src/lib.rs:54]:
```suggestion
/// Substracts an amount from a balance and reports an error if that is impossible
```

File: linera-execution/src/lib.rs [hunk 13]
```diff
@@ -50,6 +51,72 @@ use serde::{Deserialize, Serialize};
...
+pub fn sub_assign_fees(balance: &mut Amount, fees: Amount) -> Result<(), PricingError> {
+    Ok(balance
+        .try_sub_assign(fees)
+        .map_err(|_| ArithmeticError::Underflow)?)
+}
+
+/// The entries of the runtime related to fuel and storage
```
Comment (afck) [inline linera-execution/src/lib.rs:61]: It seems to be only storage, not fuel?

File: linera-execution/src/lib.rs [hunk 14]
```diff
@@ -50,6 +51,72 @@ use serde::{Deserialize, Serialize};
...
+#[derive(Copy, Debug, Clone)]
+pub struct RuntimeLimits {
+    /// The maximum size of read allowed per block
+    pub maximum_bytes_read: u64,
+    /// The maximum size of write allowed per block
+    pub maximum_bytes_written: u64,
+}
+
+impl Default for RuntimeLimits {
```
Comment (afck) [inline linera-execution/src/lib.rs:70]: Should this be test-only?

File: linera-execution/src/lib.rs [hunk 15]
```diff
@@ -50,6 +51,72 @@ use serde::{Deserialize, Serialize};
...
+#[derive(Copy, Debug, Clone)]
+pub struct RuntimeCounts {
+    /// The remaining fuel available
+    pub remaining_fuel: u64,
+    /// The number of read operations
+    pub num_reads: u64,
+    /// The bytes that have been read
+    pub bytes_read: u64,
+    /// The bytes that have been written
+    pub bytes_written: u64,
```
Comment (afck) [inline linera-execution/src/lib.rs:89]: It would be nicer if we treated all categories the same: Either we subtract from the remaining allowance, or we add to the usage so far.

I think we'll want to get to a point where execution stops as soon as it's clear that the balance would be insufficient to pay for it, so maybe the former is better.

Or should we rather track the remaining _balance_, and make the pricing available everywhere? That seems to be the only way to stop execution as soon as the _sum_ of all used resources becomes too high.

File: linera-execution/src/lib.rs [hunk 16]
```diff
@@ -50,6 +51,72 @@ use serde::{Deserialize, Serialize};
...
+) -> Result<(), ExecutionError> {
+    let initial_fuel = pricing.remaining_fuel(balance.clone());
+    let used_fuel = initial_fuel.saturating_sub(runtime_counts.remaining_fuel);
+    sub_assign_fees(balance, pricing.fuel_price(used_fuel)?)?;
+    sub_assign_fees(
+        balance,
+        pricing.storage_num_reads_price(&runtime_counts.num_reads)?,
+    )?;
+    let bytes_read = runtime_counts.bytes_read;
+    runtime_limits.maximum_bytes_read -= bytes_read;
```
Comment (afck) [inline linera-execution/src/lib.rs:106]: We should either use `checked_sub` everywhere, or add documentation to this function that the caller needs to ensure that the counts don't exceed the limits.

File: linera-execution/src/pricing.rs [hunk 17]
```diff
@@ -33,10 +56,25 @@ impl Pricing {
         Ok(self.messages.try_mul(size)?)
     }
 
-    pub fn storage_price(&self, data: &impl Serialize) -> Result<Amount, PricingError> {
+    pub fn storage_num_reads_price(&self, data: &impl Serialize) -> Result<Amount, PricingError> {
+        let size =
+            u128::try_from(bcs::serialized_size(data)?).map_err(|_| ArithmeticError::Overflow)?;
+        Ok(self.storage_num_reads.try_mul(size)?)
```
Comment (afck) [inline linera-execution/src/pricing.rs:62]: `num_reads` suggests that this is about the fixed price per read operation, not about the price per byte.

File: linera-execution/src/runtime.rs [hunk 18]
```diff
@@ -33,12 +33,22 @@ use std::{
 /// Runtime data tracked during the execution of a transaction.
 #[derive(Debug, Clone)]
 pub(crate) struct ExecutionRuntime<'a, C, const WRITABLE: bool> {
+    /// The amount of fuel available for executing the application.
+    remaining_fuel: Arc<AtomicU64>,
+    /// the number of read
```
Comment (afck) [inline linera-execution/src/runtime.rs:38]:
```suggestion
    /// The number of reads
```

File: linera-execution/src/runtime.rs [hunk 19]
```diff
@@ -33,12 +33,22 @@ use std::{
...
+    num_reads: Arc<AtomicU64>,
+    /// the total size being read
+    bytes_read: Arc<AtomicU64>,
+    /// the total size being write
```
Comment (afck) [inline linera-execution/src/runtime.rs:42]:
```suggestion
    /// The total size being written
```

(Or "that has been written so far"?)

File: linera-execution/src/runtime.rs [hunk 20]
```diff
@@ -368,7 +390,19 @@ where
...
+                    self.bytes_read
+                        .fetch_add(value.len() as u64, Ordering::Relaxed);
+                }
+                let bytes_read = self.bytes_read.load(Ordering::Acquire);
+                if bytes_read > self.maximum_bytes_read {
+                    return Err(ExecutionError::ExcessiveRead);
+                }
+                Ok(result)
+            }
             None => Err(ExecutionError::ApplicationStateNotLocked),
```
Comment (afck) [inline linera-execution/src/runtime.rs:406]: Since the `Some` arm is now quite long, you could use `ok_or_else` (or `ok_or` if execution errors are small) instead of the `match`. (Also below.)

File: linera-execution/src/runtime.rs [hunk 21]
```diff
@@ -483,6 +557,12 @@ where
     }
 
     async fn write_batch_and_unlock(&self, batch: Batch) -> Result<(), ExecutionError> {
+        let size = batch.size() as u64;
+        self.bytes_written.fetch_add(size, Ordering::Relaxed);
```
Comment (afck) [inline linera-execution/src/runtime.rs:561]: (In theory this could overflow. Probably not in practice… but maybe the default for these limits should not be `MAX` but rather `MAX / 2` or something?)

File: linera-execution/tests/wasm.rs [hunk 22]
```diff
@@ -25,6 +26,7 @@ use test_case::test_case;
 /// called correctly and consume the expected amount of fuel.
 ///
 /// To update the bytecode files, run `linera-execution/update_wasm_fixtures.sh`.
+#[ignore]
```
Comment (afck) [inline linera-execution/tests/wasm.rs:29]: (This is just temporary, is it? Because you haven't updated the numbers yet?)

File: linera-service/src/linera.rs [hunk 23]
```diff
@@ -1359,23 +1395,50 @@ impl Runnable for Job {
...
                         if let Some(messages) = messages {
                             pricing.messages = messages;
                         }
                         info!(
                             "Pricing:\n\
                             {:.2} base cost per block\n\
-                            {:.2} per unit of fuel used in executing user operations and messages\n\
+                            {:.2} per unit of fuel used in the read per operation\n\
+                            {:.2} per unit of fuel used in the read per byte\n\
+                            {:.2} per unit of fuel used in the write per byte\n\
```
Comment (afck) [inline linera-service/src/linera.rs:1421]: Do we call this "fuel" at all?
I think the first one should just be "per read operation", and the others "per byte read" and "per byte written"?

File: linera-execution/tests/wasm.rs (hunk 22 above)
Comment (MathieuDutSik) [reply linera-execution/tests/wasm.rs:29]: Yes, temporary.

File: linera-chain/src/chain.rs [hunk 24]
```diff
@@ -483,14 +484,25 @@ where
...
+        sub_assign_fees(
+            balance,
+            pricing.storage_bytes_written_price(&block.operations)?,
+        )?;
 
         let mut messages = Vec::new();
         let mut message_counts = Vec::new();
-        let available_fuel = pricing.remaining_fuel(*balance);
-        let mut remaining_fuel = available_fuel;
+        //        let available_fuel = pricing.remaining_fuel(*balance);
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-chain/src/chain.rs:499]: Yes, but the work was not finished.

File: linera-execution/src/lib.rs (hunk 14 above)
Comment (MathieuDutSik) [reply linera-execution/src/lib.rs:70]: Yes, right.

File: linera-execution/src/lib.rs (hunk 15 above)
Comment (MathieuDutSik) [reply linera-execution/src/lib.rs:89]: Yes, it would be nice, but I try to have a realistic solution here.

I tried indeed to have the execution ending as soon as possible, as I hope is clear. But the `remaining_fuel` is managed as a `u64` so I did not see a simple way to get it closer to the heart of the matter. It is done after the runtime execution.

And yes, that involves the `pricing` information being available everywhere. Since there are specific costs associated with execution and storage and each and everyone can trigger early termination, there is no way around it.

File: linera-execution/src/pricing.rs (hunk 17 above)
Comment (MathieuDutSik) [reply linera-execution/src/pricing.rs:62]: And yes, it is that. It is price per read operation, not price per byte.

File: linera-execution/src/execution.rs (hunk 7 above)
Comment (MathieuDutSik) [reply linera-execution/src/execution.rs]: This has been addressed in subsequent work of this PR.

File: linera-execution/src/lib.rs (hunk 13 above)
Comment (MathieuDutSik) [reply linera-execution/src/lib.rs:61]: Yes.

File: linera-execution/src/runtime.rs (hunk 20 above)
Comment (MathieuDutSik) [reply linera-execution/src/runtime.rs:406]: I tried that and could not find a solution easily. Since this has no functional impact, I propose not do this change.

File: linera-service/src/linera.rs (hunk 23 above)
Comment (MathieuDutSik) [reply linera-service/src/linera.rs:1421]: Yes, you are right.

File: linera-chain/src/chain.rs [hunk 25]
```diff
@@ -541,12 +558,8 @@ where
             message_counts
                 .push(u32::try_from(messages.len()).map_err(|_| ArithmeticError::Overflow)?);
         }
-        let used_fuel = available_fuel.saturating_sub(remaining_fuel);
-
         let balance = self.execution_state.system.balance.get_mut();
-        Self::sub_assign_fees(balance, credit)?;
-        Self::sub_assign_fees(balance, pricing.fuel_price(used_fuel)?)?;
-        Self::sub_assign_fees(balance, pricing.messages_price(&messages)?)?;
+        sub_assign_fees(balance, credit)?;
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-chain/src/chain.rs:561]: I am not sure what you mean by "added twice", but the logic is definitely complicated and took me some time to deal with. But I think this should be in separate PRs.

File: linera-execution/src/execution.rs (hunk 11 above)
Comment (MathieuDutSik) [reply linera-execution/src/execution.rs:250]: Yes.

File: linera-execution/src/lib.rs (hunk 16 above)
Comment (MathieuDutSik) [reply linera-execution/src/lib.rs:107]: I added the documentation.

File: linera-execution/src/pricing.rs (hunk 2 above)
Comment (MathieuDutSik) [reply linera-execution/src/pricing.rs]: Ok, I renamed it `ResourceControlPolicy`.
I also renamed `pricing` as `policy` when appropriate.
I kept the `PricingError` since it occurs only for monetary reasons.

File: linera-chain/src/chain.rs (hunk 25 above)
[the remark this answers is not in the document]
Comment (afck) [reply linera-chain/src/chain.rs:561]: `credit` is the sum of all incoming transfers in this block. To avoid failing due to insufficient balance we add this to the balance before subtracting any costs. That way, even if your current balance is 0, if you have an incoming transfer you can create a new block and receive that transfer.

Since during the actual execution of the block the transfer messages are processed, their amounts are added to the chain balance _again_. That's why, after execution, we subtract `credit` again.

With this PR, however, we use the intermediate values of `balance` to compute our fuel limit, too, and in these, the incoming transfers will sometimes already be counted twice.

To fix this, instead of subtracting `credit` again at the end, we could make the execution of a transfer a no-op. Not really elegant either, though. :confused:

File: linera-execution/src/runtime.rs (hunk 20 above)
Comment (afck) [reply linera-execution/src/runtime.rs:406]: Doesn't `.ok_or_else(|| ExecutionError::ApplicationStateNotLocked)?` work?

File: linera-execution/tests/wasm.rs (hunk 22 above)
Comment (MathieuDutSik) [reply linera-execution/tests/wasm.rs]: Addressed.

File: linera-execution/src/runtime.rs (hunk 20 above)
Comment (MathieuDutSik) [reply linera-execution/src/runtime.rs:442]: No.

File: linera-execution/src/lib.rs (hunk 15 above)
Comment (MathieuDutSik) [reply linera-execution/src/lib.rs:117]: But knowing the maximum value for fuel, bytes_written, num_reads, and bytes_read does not mean that we know how much should occur. Because if every maximum values is attained then it does not work.

File: linera-execution/src/execution.rs (hunk 8 above)
Comment (afck) [reply linera-execution/src/execution.rs]: (Addressing this in https://github.com/linera-io/linera-protocol/pull/1160.)

File: linera-execution/src/lib.rs (hunk 15 above)
Comment (afck) [reply linera-execution/src/lib.rs:117]: Yes, you could still cause four times as much resource usage as your balance allows.
But that's still better than 100 or 1000 times, which is entirely possible if we only go by the fixed maximum values.

File: linera-execution/src/pricing.rs (hunk 17 above)
Comment (afck) [reply linera-execution/src/pricing.rs:62]: But you're still multiplying it with the size in bytes here.

File: linera-execution/src/lib.rs (hunk 15 above)
Comment (MathieuDutSik) [reply linera-execution/src/lib.rs:117]: Ok, if that is what you want, then so be it.

File: linera-execution/src/pricing.rs [hunk 26]
```diff
@@ -33,10 +56,22 @@ impl Pricing {
         Ok(self.messages.try_mul(size)?)
     }
 
-    pub fn storage_price(&self, data: &impl Serialize) -> Result<Amount, PricingError> {
+    pub fn storage_n_read_price(&self, data: &impl Serialize) -> Result<Amount, PricingError> {
+        let size =
+            u128::try_from(bcs::serialized_size(data)?).map_err(|_| ArithmeticError::Overflow)?;
+        Ok(self.storage_n_read.try_mul(size)?)
+    }
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-execution/src/pricing.rs]: Yes, corrected.

File: linera-execution/src/pricing.rs (hunk 17 above)
Comment (MathieuDutSik) [reply linera-execution/src/pricing.rs:62]: Ok, corrected.

File: linera-execution/src/lib.rs [hunk 27]
```diff
@@ -50,6 +50,46 @@ use serde::{Deserialize, Serialize};
...
+    /// The number of read operations
+    pub n_read: u64,
+    /// The bytes that have been read
+    pub bytes_read: u64,
+    /// The bytes that have been written
+    pub bytes_write: u64,
+    /// The maximum size of read allowed
+    pub maximum_bytes_read: u64,
+    /// The maximum size of write allowed
+    pub maximum_bytes_write: u64,
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-execution/src/lib.rs]: So, this has been renamed as `num_reads` and the structure has been significantly changed.

File: linera-chain/src/chain.rs (hunk 6 above)
Comment (MathieuDutSik) [reply linera-chain/src/chain.rs]: The block_storage price is something we want to implement afterward.

File: linera-execution/src/runtime.rs (hunk 3 above)
Comment (MathieuDutSik) [reply linera-execution/src/runtime.rs]: I did change the structure of the code and added an `update_limits` method to the `RuntimeLimits`.
But the structure has changed a lot now.

File: linera-chain/src/chain.rs [hunk 28]
```diff
@@ -494,14 +495,28 @@ where
...
+            policy.storage_bytes_written_price_raw(&block.operations)?,
+        )?;
 
         let mut messages = Vec::new();
         let mut message_counts = Vec::new();
-        let available_fuel = pricing.remaining_fuel(*balance);
-        let mut remaining_fuel = available_fuel;
+        let maximum_bytes_read = policy.maximum_bytes_read;
+        let maximum_bytes_written = policy.maximum_bytes_written;
+        let mut runtime_track = RuntimeTracker {
```
Comment (ma2bd) [inline linera-chain/src/chain.rs:512]: `tracker` or `runtime_tracker` please

File: linera-chain/src/chain.rs [hunk 29]
```diff
@@ -494,14 +495,28 @@ where
...
+            used_fuel: 0,
+            num_reads: 0,
+            bytes_read: 0,
+            bytes_written: 0,
+            maximum_bytes_read,
```
Comment (ma2bd) [inline linera-chain/src/chain.rs:517]: Why copy elements of the policy inside the tracker?

File: linera-chain/src/chain.rs [hunk 30]
```diff
@@ -494,14 +495,28 @@ where
         let balance = self.execution_state.system.balance.get_mut();
 
         balance.try_add_assign(credit)?;
-        Self::sub_assign_fees(balance, pricing.certificate_price())?;
-        Self::sub_assign_fees(balance, pricing.storage_price(&block.incoming_messages)?)?;
-        Self::sub_assign_fees(balance, pricing.storage_price(&block.operations)?)?;
+        sub_assign_fees(balance, policy.certificate_price())?;
```
Comment (ma2bd) [inline linera-chain/src/chain.rs:498]: Removing Self:: may or may not be an improvement.

File: linera-chain/src/chain.rs (hunk 25 above)
[the remark this answers is not in the document]
Comment (ma2bd) [reply linera-chain/src/chain.rs:576]: Did the internal discussion clarify this? or do we need to coordinate?

File: linera-execution/src/lib.rs [hunk 31]
```diff
@@ -50,6 +51,136 @@ use serde::{Deserialize, Serialize};
 use std::{io, path::Path, str::FromStr, sync::Arc};
 use thiserror::Error;
 
+/// Substracts an amount from a balance and reports an error if that is impossible
+pub fn sub_assign_fees(balance: &mut Amount, fees: Amount) -> Result<(), PricingError> {
```
Comment (ma2bd) [inline linera-execution/src/lib.rs:55]: ok so for a public APIs in general I try to use methods because explicitly importing each public function is tedious.

File: linera-execution/src/lib.rs [hunk 32]
```diff
@@ -50,6 +51,136 @@ use serde::{Deserialize, Serialize};
...
+    pub max_budget_bytes_written: u64,
+    /// The maximum size of read allowed per block
+    pub maximum_bytes_read: u64,
+    /// The maximum size of write allowed per block
+    pub maximum_bytes_written: u64,
+}
+
+/// The entries of the runtime related to storage
+#[derive(Copy, Debug, Clone)]
+pub struct RuntimeTracker {
```
Comment (ma2bd) [inline linera-execution/src/lib.rs:78]: Would ResourceTracker make sense?

File: linera-execution/src/runtime.rs [hunk 33]
```diff
@@ -33,12 +33,20 @@ use std::{
 /// Runtime data tracked during the execution of a transaction.
 #[derive(Debug, Clone)]
 pub(crate) struct ExecutionRuntime<'a, C, const WRITABLE: bool> {
+    /// The amount of fuel available for executing the application.
```
Comment (ma2bd) [inline linera-execution/src/runtime.rs:36]: Can we group the resource fields? Also I would generally add new fields at the end if there is no obvious choice

File: linera-chain/src/chain.rs (hunk 25 above)
[the remark this answers is not in the document]
Comment (afck) [reply linera-chain/src/chain.rs:576]: To summarize the discussion:
* Let's remove the `credit` thing: If you have 0 balance, you have to make sure that the first incoming message is a transfer that covers the block's fees.
* Subtract the base fee for the block _after_ processing messages and operations.
* Remove the fee for storing incoming messages. Consider this to be included in the `messages_price` on the sender chain.

[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-chain/src/chain.rs:576]: Ok, I would propose to do that after in another PR.

File: linera-chain/src/chain.rs (hunk 28 above)
Comment (MathieuDutSik) [reply linera-chain/src/chain.rs:512]: Ok.

File: linera-execution/src/lib.rs (hunk 32 above)
Comment (MathieuDutSik) [reply linera-execution/src/lib.rs:78]: Yes, it does.

File: linera-execution/src/runtime.rs (hunk 33 above)
Comment (MathieuDutSik) [reply linera-execution/src/runtime.rs:36]: Bu grouping, do you mean replace the `Arc<Atomic...>` by `Arc<Mutes<RuntimeCounts>>`?

File: linera-chain/src/chain.rs (hunk 29 above)
Comment (ma2bd) [reply linera-chain/src/chain.rs:517]: I see. Perhaps different names could help. (We're conflating the policy `maximum_bytes_read_per_block` with the tracker `maximum_bytes_left_to_read`.)

File: linera-execution/src/lib.rs [hunk 34]
```diff
@@ -50,6 +51,136 @@ use serde::{Deserialize, Serialize};
...
+            max_budget_bytes_read: u64::MAX / 2,
+            max_budget_bytes_written: u64::MAX / 2,
+            maximum_bytes_read: u64::MAX / 2,
+            maximum_bytes_written: u64::MAX / 2,
+        }
+    }
+}
+
+impl ResourceTracker {
+    /// Update the limits for the maximum and update the balance.
```
Comment (afck) [inline linera-execution/src/lib.rs:120]:
```suggestion
    /// Updates the limits for the maximum and updates the balance.
```

File: linera-execution/src/lib.rs [hunk 35]
```diff
@@ -50,6 +51,136 @@ use serde::{Deserialize, Serialize};
...
+    /// The substraction are guaranteed to work correctly and the
```
Comment (afck) [inline linera-execution/src/lib.rs:121]:
```suggestion
    /// The substractions are guaranteed to work correctly and the
```

(But actually I don't think that needs to be in the method documentation. It would only have to be there if they were _not_ guaranteed to work correctly.)

File: linera-execution/src/lib.rs [hunk 36]
```diff
@@ -50,6 +51,136 @@ use serde::{Deserialize, Serialize};
...
+        // The number of reads
+        sub_assign_fees(
+            balance,
+            policy.storage_num_reads_price(&runtime_counts.num_reads)?,
+        )?;
+        self.num_reads += runtime_counts.num_reads;
+        // The number of bytes read
+        let bytes_read = runtime_counts.bytes_read;
+        self.maximum_bytes_read -= bytes_read;
+        self.bytes_read += runtime_counts.bytes_read;
```
Comment (afck) [inline linera-execution/src/lib.rs:143]: Is it correct to both remove them from the maximum _and_ add them to `bytes_read`?

File: linera-execution/src/policy.rs [hunk 37]
```diff
@@ -33,10 +56,28 @@ impl Pricing {
         Ok(self.messages.try_mul(size)?)
     }
 
-    pub fn storage_price(&self, data: &impl Serialize) -> Result<Amount, PricingError> {
+    pub fn storage_num_reads_price(&self, size: &u64) -> Result<Amount, PricingError> {
+        let size = *size as u128;
+        Ok(self.storage_num_reads.try_mul(size)?)
```
Comment (afck) [inline linera-execution/src/policy.rs:61]: Is this still being multiplied by size? Or does "size" mean "number of operations" here, as opposed to in the other methods? Then I'd call it "count".

File: linera-execution/src/runtime.rs [hunk 38]
```diff
@@ -445,6 +528,19 @@ where
         self.remaining_fuel.load(Ordering::Acquire)
     }
 
+    fn runtime_counts(&self) -> RuntimeCounts {
+        let remaining_fuel = self.remaining_fuel.load(Ordering::Acquire);
+        let num_reads = self.num_reads.load(Ordering::Acquire);
```
Comment (afck) [inline linera-execution/src/runtime.rs:533]: I forgot… why is there only a `num_reads` but no `num_writes`?

File: linera-chain/src/chain.rs (hunk 29 above)
Comment (MathieuDutSik) [reply linera-chain/src/chain.rs]: Thanks, renamed.

File: linera-execution/src/runtime.rs (hunk 38 above)
Comment (MathieuDutSik) [reply linera-execution/src/runtime.rs:533]: It is because we have a single write at the end of the block when the `.save()` is called.

File: linera-execution/src/policy.rs (hunk 37 above)
Comment (MathieuDutSik) [reply linera-execution/src/policy.rs:61]: Then I renamed as `count`. Thanks.

File: linera-execution/src/lib.rs (hunk 36 above)
Comment (MathieuDutSik) [reply linera-execution/src/lib.rs:143]: Yes, because the `bytes_read` as a control variable that could be used for `Prometheus`.

File: linera-execution/src/runtime.rs (hunk 20 above)
Comment (MathieuDutSik) [reply linera-execution/src/runtime.rs:437]: I could find a way to address the livedness issues.

File: linera-execution/src/lib.rs (hunk 15 above)
Comment (MathieuDutSik) [reply linera-execution/src/lib.rs:179]: So, this has been implemented.

File: linera-chain/src/chain.rs (hunk 25 above)
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-chain/src/chain.rs:571]: I created the corresponding issue https://github.com/linera-io/linera-protocol/issues/1187

File: linera-execution/src/lib.rs (hunk 31 above)
Comment (MathieuDutSik) [reply linera-execution/src/lib.rs:55]: So, far this is a single import. We can think of something else in the future if that grows out.

File: linera-chain/src/chain.rs (hunk 30 above)
Comment (MathieuDutSik) [reply linera-chain/src/chain.rs:498]: It may or it may not, I agree, but I see this as rather not essential.

**Review by afck — APPROVED** [review]
